### PR TITLE
descpb,descbuilder: clean up descriptor timestamp logic

### PR DESCRIPTION
**Pull request by postamar**
Previously, creating a descriptor builder from a protobuf could trigger
a panic when the descriptor's ModificationTime and Version fields were
inconsistent with the provided MVCC Timestamp.

This logic dates back to before the existence of descriptor builders or
the unification of descriptor reads from storage and had to assume that
any descriptor protobuf may have been freshly read from storage.

These panics are better than severely corrupting the schema but are
otherwise quite undesirable. This commit moves this ModificationTime
update logic to the builder's RunPostDeserializationChanges method
because that's where it belongs.

The challenge is that post-deserialization changes have to be idempotent,
also an MVCC timestamp is not always available because builders get
created from protobufs originating from various sources other than
a roachpb.Value read from system.descriptors:
  - backup manifests,
  - changefeeds,
  - existing catalog.Descriptor instances,
  - etc.

Release note: None

### CODE REVIEW

**Comment by postamar** [comment]
This is ready for review. There are still a couple of test failures but I know why they're occurring and I strongly believe they're peripheral to the change I made. 

I recommend reviewers look at the changes in `descpb` and `descbuilder` first. This commit clocks in at 1k+ lines but most of these are noise.

**Comment by Xiang-Gu** [comment]
so, it seems to me what you did is:
1. Add a time field in the descriptor builder that stores the possible mvcc timestamp read from storage;
2. Once we have a descriptor builder, we can now modify its `maybeModified` catalog.descriptor through `postdeserializationchanges`, and you move the logic of set the modificationTime field from the mvccTimestamp there. That's why you need step 1 to find the mvccTimestamp a "home" to live in  -- that home is the descriptor builder;


Of course, there are cases where we construct a descriptor builder without having a mvccTimestamp, like you listed. Then the builder will carry an empty timestamp. How do we handle those situations? How did you tackle the challenges, as you mentioned
> The challenge is that post-deserialization changes have to be idempotent,
also an MVCC timestamp is not always available because builders get
created from protobufs originating from various sources other than
a roachpb.Value read from system.descriptors:

File: pkg/sql/catalog/dbdesc/database_desc_builder.go [hunk 1]
```diff
@@ -68,8 +87,21 @@ func (ddb *databaseDescriptorBuilder) DescriptorType() catalog.DescriptorType {
 
 // RunPostDeserializationChanges implements the catalog.DescriptorBuilder
 // interface.
-func (ddb *databaseDescriptorBuilder) RunPostDeserializationChanges() error {
+func (ddb *databaseDescriptorBuilder) RunPostDeserializationChanges() (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "database %q (%d)", ddb.original.Name, ddb.original.ID)
+	}()
+	mustSetModTime, err := descpb.MustSetModificationTime(
+		ddb.original.ModificationTime, ddb.mvccTimestamp, ddb.original.Version,
```
[the remark this answers is not in the document]
Comment (postamar) [reply pkg/sql/catalog/dbdesc/database_desc_builder.go:95]: I wanted to stress the fact that ModificationTime needed to be set before doing any other change. I'll add a comment to that effect to explain this.

**Comment by postamar** [comment]
> How do we handle those situations?

We don't. Passing an empty MVCC timestamp along with a protobuf to the builder constructor is only valid when we know that the protobuf originates from an existing `catalog.Descriptor` which has already been subjected to post-deserialization changes..

**Review by Xiang-Gu — APPROVED** [review]
LGTM! My #88844 will wait and use this PR to pass `rawBytesInStorage` into the builder constructor.

CI are all "flaky" tests and I've re-run it.

File: pkg/sql/catalog/descbuilder/desc_builder.go [hunk 2]
```diff
@@ -21,39 +22,151 @@ import (
...
+		}
+		if original == nil {
+			*newVersion = 1
+		} else if *newVersion != original.GetVersion() {
+			*newVersion = original.GetVersion()
+		}
+		if !mvccTimestamp.IsEmpty() && newModTime.IsEmpty() {
+			*newModTime = mvccTimestamp
+		}
+	}
```
Comment (chengxiong-ruan) [inline pkg/sql/catalog/descbuilder/desc_builder.go:126]: I trust you on this. But I'd like to have some explanation why this kind correction is ok. Specifically, how come descriptor ID can be different and why that is "fine"?

**Comment by postamar** [comment]
Thanks for the reviews!

bors r+

**Pull request merged**